### PR TITLE
Python generated from SMOP

### PR DESCRIPTION
**Pull request by dorchard**
This comprises the automatically generate code from @tztsai 's augmented SMOP translator. Some of this can be used to provide the new Python translation, perhaps with then extra documentation added and some parts made into more idiomatic Python (either by hand or via an LLM) as time permits.

### CODE REVIEW

**Comment by dorchard** [comment]
I think this can be merged now. Some of the implementations can be used.

**Review by j-emberton — DISMISSED** [review]
This looks ok to me. I like the addition of the `ruff` pre commit hook. All tests pass.

**Comment by dorchard** [comment]
Tests all pass now

File: tests/unit_tests/test_ustar_cp/test_cpdBootstrapUStarTh4Season20100901.py [hunk 1]
```diff
@@ -261,12 +261,13 @@ def test_get_itNee(test_engine, NEE, uStar, T, iNight, expected_itNee):
...
     # Compare results
-    if type(itNee) != float:
+    if not isinstance(itNee, float):
         assert compare_matlab_arrays(itNee, expected_itNee), f"Expected {expected_itNee}, but got {itNee}"
     else:
-        assert itNee==expected_itNee
+        assert np.allclose(itNee, expected_itNee)
 
 # Test for the setup_Cp function
+@pytest.mark.skip(reason="Too long")
```
Comment (j-emberton) [inline tests/unit_tests/test_ustar_cp/test_cpdBootstrapUStarTh4Season20100901.py]:
```suggestion

```

**Pull request merged**